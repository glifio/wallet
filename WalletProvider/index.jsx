import { createContext, useContext, useReducer, useCallback } from 'react'
import PropTypes from 'prop-types'

import reducer, {
  initialState,
  setWalletType,
  setError,
  resetLedgerState,
  resetState
} from './state'
import { setLedgerProvider } from '../utils/ledger/setLedgerProvider'
import fetchDefaultWallet from './fetchDefaultWallet'
import connectLedger from './connectLedger'
import { useWasm } from '../lib/WasmLoader'

export const WalletProviderContext = createContext({})

const WalletProviderWrapper = ({ network, children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)
  const { walletSubproviders } = useWasm()
  return (
    <WalletProviderContext.Provider
      value={{
        state,
        dispatch,
        fetchDefaultWallet: useCallback(
          // a lot of times here, we instantiate the walletProvider AND fetch the default wallet back-to-back
          // which could lead to race conditions, since the wallet provider's state may not have updated in time
          // thats why we allow you to pass the walletProvider here, and fallback to the provider in state in other circumstances
          (walletProvider = state.walletProvider) =>
            fetchDefaultWallet(
              dispatch,
              network,
              state.walletType,
              walletProvider,
              walletSubproviders
            ),
          [
            dispatch,
            network,
            state.walletType,
            state.walletProvider,
            walletSubproviders
          ]
        ),
        setWalletError: errorMessage => dispatch(setError(errorMessage)),
        setWalletType: walletType => dispatch(setWalletType(walletType)),
        setLedgerProvider: useCallback(
          () => setLedgerProvider(dispatch, walletSubproviders.LedgerProvider),
          [dispatch, walletSubproviders.LedgerProvider]
        ),
        connectLedger: useCallback(
          () => connectLedger(dispatch, walletSubproviders.LedgerProvider),
          [dispatch, walletSubproviders.LedgerProvider]
        ),
        resetLedgerState: () => dispatch(resetLedgerState()),
        resetState: useCallback(() => dispatch(resetState()), [dispatch]),
        walletSubproviders
      }}
    >
      {children}
    </WalletProviderContext.Provider>
  )
}

WalletProviderWrapper.propTypes = {
  children: PropTypes.node.isRequired,
  network: PropTypes.oneOf(['t', 'f']).isRequired
}

export const useWalletProvider = () => {
  const value = useContext(WalletProviderContext)
  const { state } = value
  return {
    ...state,
    ...value
  }
}

export default WalletProviderWrapper
