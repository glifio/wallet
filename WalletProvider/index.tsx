import {
  createContext,
  useContext,
  useReducer,
  useCallback,
  Dispatch
} from 'react'
import PropTypes from 'prop-types'
import Filecoin, { LedgerProvider } from '@glif/filecoin-wallet-provider'

import reducer, {
  initialState,
  setLoginOption,
  setError,
  resetLedgerState,
  resetState,
  walletList,
  switchWallet,
  updateBalance
} from './state'
import fetchDefaultWallet from './fetchDefaultWallet'
import connectLedger from './connectLedger'
import {
  LoginOption,
  Wallet,
  WalletProviderAction,
  WalletProviderState
} from './types'
import { FilecoinNumber } from '@glif/filecoin-number'

/* eslint-disable no-unused-vars */
type WalletProviderContextType = {
  state: WalletProviderState
  dispatch: Dispatch<WalletProviderAction> | null
  fetchDefaultWallet: (walletProvider: Filecoin) => Promise<Wallet>
  connectLedger: () => Promise<Filecoin>
  setWalletError: (errorMessage: string) => void
  setLoginOption: (loginOption: LoginOption) => void
  resetLedgerState: () => void
  resetState: () => void
  walletList: (wallets: Wallet[]) => void
  switchWallet: (walletIdx: number) => void
  updateBalance: (bal: FilecoinNumber, walletIdx: number) => void
}

export const WalletProviderContext = createContext<WalletProviderContextType>({
  state: { ...initialState },
  dispatch: null,
  fetchDefaultWallet: null,
  connectLedger: null,
  setWalletError: null,
  setLoginOption: null,
  resetLedgerState: null,
  resetState: null,
  walletList: null,
  switchWallet: null,
  updateBalance: null
})

const WalletProviderWrapper = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)
  return (
    <WalletProviderContext.Provider
      value={{
        state,
        dispatch,
        fetchDefaultWallet: useCallback(
          // a lot of times here, we instantiate the walletProvider AND fetch the default wallet back-to-back
          // which could lead to race conditions, since the wallet provider's state may not have updated in time
          // thats why we allow you to pass the walletProvider here, and fallback to the provider in state in other circumstances
          (walletProvider = state.walletProvider) =>
            fetchDefaultWallet(dispatch, state.loginOption, walletProvider),
          [dispatch, state.loginOption, state.walletProvider]
        ),
        setWalletError: useCallback(
          (errorMessage) => dispatch(setError(errorMessage)),
          [dispatch]
        ),
        setLoginOption: useCallback(
          (loginOption) => dispatch(setLoginOption(loginOption)),
          [dispatch]
        ),
        connectLedger: useCallback(
          () =>
            connectLedger(
              dispatch,
              state?.walletProvider?.wallet as LedgerProvider
            ),
          [dispatch, state?.walletProvider?.wallet]
        ),
        resetLedgerState: useCallback(
          () => dispatch(resetLedgerState()),
          [dispatch]
        ),
        resetState: useCallback(() => dispatch(resetState()), [dispatch]),
        walletList: useCallback(
          (wallets) => dispatch(walletList(wallets)),
          [dispatch]
        ),
        switchWallet: useCallback(
          (selectedWalletIdx) => dispatch(switchWallet(selectedWalletIdx)),
          [dispatch]
        ),
        updateBalance: useCallback(
          (balance, index) => dispatch(updateBalance(balance, index)),
          [dispatch]
        )
      }}
    >
      {children}
    </WalletProviderContext.Provider>
  )
}

WalletProviderWrapper.propTypes = {
  children: PropTypes.node.isRequired
}

export const useWalletProvider = () => {
  const value = useContext(WalletProviderContext)
  const { state } = value
  return {
    ...state,
    ...value
  }
}

export default WalletProviderWrapper
