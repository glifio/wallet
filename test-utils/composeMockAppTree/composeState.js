import cloneDeep from 'lodash.clonedeep'
import { FilecoinNumber } from '@openworklabs/filecoin-number'

import createPath from '../../utils/createPath'
import { IMPORT_MNEMONIC } from '../../constants'
import { initialState } from '../../store/states'
import { mockWalletProviderInstance } from '../mocks/mock-wallet-provider'

export const presets = {
  preOnboard: cloneDeep(initialState),
  postOnboard: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi',
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  }),
  postOnboardWithError: cloneDeep({
    ...initialState,
    error: 'error for testing',
    wallets: [
      {
        address: 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi',
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  })
}

export const composeWalletProviderState = (
  initialWalletProviderState,
  preset
) => {
  switch (preset) {
    case 'postOnboard': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance
      })
    }
    case 'postOnboardWithError': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance
      })
    }
    default:
      return initialWalletProviderState
  }
}
