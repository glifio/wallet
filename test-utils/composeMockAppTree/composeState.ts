import cloneDeep from 'lodash.clonedeep'
import { FilecoinNumber } from '@glif/filecoin-number'

import createPath from '../../utils/createPath'
import { IMPORT_MNEMONIC } from '../../constants'
import { initialState } from '../../store/states'
import { mockWalletProviderInstance } from '../mocks/mock-wallet-provider'
import { emptyMsigState } from '../../MsigProvider/types'
import { WALLET_ADDRESS, MULTISIG_ACTOR_ADDRESS, signers } from '../constants'

export const presets = {
  preOnboard: cloneDeep(initialState),
  postOnboard: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: WALLET_ADDRESS,
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  }),
  postOnboardLowBal: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: WALLET_ADDRESS,
        balance: new FilecoinNumber('.000001', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  }),
  postOnboardWithError: cloneDeep({
    ...initialState,
    error: 'error for testing',
    wallets: [
      {
        address: WALLET_ADDRESS,
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  }),
  pendingMsigCreate: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: WALLET_ADDRESS,
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0,
    messages: {
      confirmed: [],
      pending: [
        {
          cid: 'bafy2bzaced2godar6dht6ag6omfa53jr6tykygtzskzodqo2k4ngvyou4yct2',
          from: 'f1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a',
          gasFeeCap: '100425',
          gasLimit: 11452096,
          gasPremium: '99371',
          maxFee: '1150076740800',
          method: 'EXEC',
          nonce: 56,
          paidFee: '0',
          params: {
            num_approvals_threshold: 1,
            signers: ['f1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a'],
            start_epoch: '0',
            unlock_duration: '0'
          },
          timestamp: 1632317830,
          to: 'f01'
        }
      ]
    }
  })
}

export const composeWalletProviderState = initialWalletProviderState => {
  return Object.freeze({
    ...initialWalletProviderState,
    walletType: IMPORT_MNEMONIC,
    walletProvider: mockWalletProviderInstance
  })
}

export const mockMsigProviderContext = {
  Address: MULTISIG_ACTOR_ADDRESS,
  ActorCode: 'fil/5/multisig',
  Balance: new FilecoinNumber('1', 'fil'),
  AvailableBalance: new FilecoinNumber('1', 'fil'),
  Signers: signers,
  InitialBalance: new FilecoinNumber('1', 'fil'),
  NextTxnID: 0,
  NumApprovalsThreshold: 1,
  StartEpoch: 0,
  UnlockDuration: 0,
  errors: {
    notMsigActor: false,
    connectedWalletNotMsigSigner: false,
    actorNotFound: false,
    unhandledError: ''
  },
  loading: false,
  setMsigActor: null
}

export const composeMsigProviderState = (preset: keyof typeof presets) => {
  switch (preset) {
    case 'preOnboard': {
      return Object.freeze(emptyMsigState)
    }
    case 'pendingMsigCreate': {
      return Object.freeze(emptyMsigState)
    }
    default: {
      return Object.freeze(mockMsigProviderContext)
    }
  }
}
