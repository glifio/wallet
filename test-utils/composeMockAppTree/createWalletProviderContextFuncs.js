import { FilecoinNumber } from '@glif/filecoin-number'

import {
  setWalletType,
  setError,
  resetLedgerState,
  resetState
} from '../../WalletProvider/state'

export default walletProviderDispatch => ({
  fetchDefaultWallet: jest.fn().mockImplementation(() => ({
    balance: new FilecoinNumber('1', 'fil'),
    address: 't1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq',
    path: "m/44'/1'/0/0/0"
  })),
  setWalletError: jest
    .fn()
    .mockImplementation(errorMessage =>
      walletProviderDispatch(setError(errorMessage))
    ),
  setWalletType: jest
    .fn()
    .mockImplementation(walletType =>
      walletProviderDispatch(setWalletType(walletType))
    ),
  setLedgerProvider: jest.fn(),
  connectLedger: jest.fn(),
  resetLedgerState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetLedgerState())
  }),
  resetState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetState())
  })
})
