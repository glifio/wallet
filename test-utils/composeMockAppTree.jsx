/* eslint-disable react/prop-types */
import React, { useReducer } from 'react'
import cloneDeep from 'lodash.clonedeep'

import WalletProvider from '@openworklabs/filecoin-wallet-provider'
import { FilecoinNumber } from '@openworklabs/filecoin-number'
import { Provider } from 'react-redux'

import { NetworkCheck } from '../lib/check-network'
import BalancePoller from '../lib/update-balance'
import { ConverterContext } from '../lib/Converter'
import { theme, ThemeProvider } from '../components/Shared'
import { initializeStore } from './index'
import { initialState } from '../store/states'
import walletProviderReducer, {
  initialState as walletProviderInitialState,
  setWalletType,
  setError,
  resetLedgerState,
  resetState
} from '../WalletProvider/state'
import { WalletProviderContext } from '../WalletProvider'
import createPath from '../utils/createPath'
import { IMPORT_MNEMONIC } from '../constants'

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

const mockRouter = jest.fn(() => {})
useRouter.mockImplementation(() => ({
  query: 'network=t',
  push: mockRouter,
  replace: mockRouter
}))

jest.mock('@openworklabs/filecoin-wallet-provider')
const mockGetAccounts = jest
  .fn()
  .mockImplementation(() =>
    Promise.resolve(['t1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq'])
  )

const mockGetBalance = jest
  .fn()
  .mockImplementation(() => Promise.resolve(new FilecoinNumber('1', 'fil')))

WalletProvider.mockImplementation(() => {
  return {
    wallet: {
      getAccounts: mockGetAccounts,
      sign: jest.fn().mockImplementation(() => 'xxxyyyyzzzz')
    },
    getBalance: mockGetBalance,
    getNonce: jest.fn().mockImplementation(() => 0),
    estimateGas: jest
      .fn()
      .mockImplementation(() => new FilecoinNumber('122', 'attofil')),
    sendMessage: jest.fn().mockImplementation(() => 'QmZCid!')
  }
})

const mockWalletProviderInstance = new WalletProvider()

const presets = {
  preOnboard: cloneDeep(initialState),
  postOnboard: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi',
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  })
}

const composeWalletProviderState = (initialWalletProviderState, preset) => {
  switch (preset) {
    case 'postOnboard': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance
      })
    }
    default:
      return initialWalletProviderState
  }
}

/**
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * pathname
 * query (usually { network: 'f' } or { network: 't' })
 */

export default (statePreset = 'preOnboard', options = {}) => {
  const state = options.state || presets[statePreset]
  const store = initializeStore(state)

  const mockConverterInstance = options.mockConverterInstance || {
    cacheConversionRate: jest.fn(),
    toFIL: jest.fn().mockImplementation(amount => {
      return new FilecoinNumber(amount, 'fil').dividedBy(5)
    }),
    fromFIL: jest.fn().mockImplementation(amount => {
      return new FilecoinNumber(amount, 'fil').multipliedBy(5)
    })
  }

  const pathname = options.pathname || '/wallet'
  const query = options.query || { network: 't' }

  const Tree = ({ children }) => {
    const [initialWalletProviderState, walletProviderDispatch] = useReducer(
      options.reducer || walletProviderReducer,
      options.walletProviderInitialState || walletProviderInitialState
    )

    const walletProviderState = composeWalletProviderState(
      initialWalletProviderState,
      statePreset
    )
    return (
      <Provider store={store}>
        <ConverterContext.Provider
          value={{
            converter: mockConverterInstance,
            converterError: options.converterError || null
          }}
        >
          <WalletProviderContext.Provider
            value={{
              state: walletProviderState,
              dispatch: walletProviderDispatch,
              fetchDefaultWallet: jest.fn().mockImplementation(),
              setWalletError: jest
                .fn()
                .mockImplementation(errorMessage =>
                  walletProviderDispatch(setError(errorMessage))
                ),
              setWalletType: jest
                .fn()
                .mockImplementation(walletType =>
                  walletProviderDispatch(setWalletType(walletType))
                ),
              setLedgerProvider: jest.fn(),
              connectLedger: jest.fn(),
              resetLedgerState: jest.fn().mockImplementation(() => {
                walletProviderDispatch(resetLedgerState())
              }),
              resetState: jest.fn().mockImplementation(() => {
                walletProviderDispatch(resetState())
              })
            }}
          >
            <NetworkCheck
              networkFromRdx={store.getState().network}
              pathname={pathname}
              query={query}
              switchNetwork={jest.fn()}
            />
            <BalancePoller />
            <ThemeProvider theme={theme}>{children}</ThemeProvider>
          </WalletProviderContext.Provider>
        </ConverterContext.Provider>
      </Provider>
    )
  }

  return { Tree, store, walletProvider: mockWalletProviderInstance }
}
