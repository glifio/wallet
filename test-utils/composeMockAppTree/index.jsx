/* eslint-disable react/prop-types */
import React, { useReducer } from 'react'
import { Provider } from 'react-redux'
import { Converter } from '@openworklabs/filecoin-number'

import { NetworkCheck } from '../../lib/check-network'
import BalancePoller from '../../lib/update-balance'
import { ConverterContext } from '../../lib/Converter'
import { theme, ThemeProvider } from '../../components/Shared'
import { initializeStore } from '..'
import walletProviderReducer, {
  initialState as walletProviderInitialState
} from '../../WalletProvider/state'
import { WasmContext } from '../../lib/WasmLoader'
import { WalletProviderContext } from '../../WalletProvider'
import { mockWalletProviderInstance } from '../mocks/mock-wallet-provider'
import createMockWalletProviderContextFuncs from './createWalletProviderContextFuncs.js'
import mockWalletSubproviders from '../mocks/mock-wallet-subproviders'
import { presets, composeWalletProviderState } from './composeState'

/**
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * walletProviderDispatch(a dispatcher to mock/read calls to the walletProvider's state)
 * pathname
 * query (usually { network: 'f' } or { network: 't' })
 */

export default (statePreset = 'preOnboard', options = {}) => {
  const state = options.state || presets[statePreset]
  const store = initializeStore(state)

  const pathname = options.pathname || '/wallet'
  const query = options.query || { network: 't' }

  const Tree = ({ children }) => {
    const [initialWalletProviderState, walletProviderDispatch] = useReducer(
      options.reducer || walletProviderReducer,
      options.walletProviderInitialState || walletProviderInitialState
    )

    const walletProviderState = composeWalletProviderState(
      initialWalletProviderState,
      statePreset
    )

    const mockWalletProviderContextFuncs = createMockWalletProviderContextFuncs(
      options.walletProviderDispatch || walletProviderDispatch
    )

    return (
      <Provider store={store}>
        <WasmContext.Provider
          value={{
            generateMnemonic: jest
              .fn()
              .mockImplementation(
                () =>
                  'slender spread awkward chicken noise useful thank dentist tip bronze ritual explain version spot collect whisper glow peanut bus local country album punch frown'
              )
          }}
        >
          <ConverterContext.Provider
            value={{
              converter: new Converter(),
              converterError: options.converterError || null
            }}
          >
            <WalletProviderContext.Provider
              value={{
                state: walletProviderState,
                dispatch:
                  options.walletProviderDispatch || walletProviderDispatch,
                walletSubproviders: {
                  ...mockWalletSubproviders
                },
                ...mockWalletProviderContextFuncs
              }}
            >
              <NetworkCheck
                networkFromRdx={store.getState().network}
                pathname={pathname}
                query={query}
                switchNetwork={jest.fn()}
              />
              <BalancePoller />
              <ThemeProvider theme={theme}>{children}</ThemeProvider>
            </WalletProviderContext.Provider>
          </ConverterContext.Provider>
        </WasmContext.Provider>
      </Provider>
    )
  }

  return {
    Tree,
    store,
    walletProvider: mockWalletProviderInstance
  }
}
