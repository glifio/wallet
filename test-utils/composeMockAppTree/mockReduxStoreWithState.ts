import { presets } from './composeState'
import { initializeStore } from '..'

export default function mockReduxStoreWithState(options: {
  state?: object
  statePreset?: string
}) {
  const state = options.state || presets[options.statePreset]
  return initializeStore(state)
}
