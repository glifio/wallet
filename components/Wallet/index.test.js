import { cleanup, render, act, fireEvent } from '@testing-library/react'
import WalletProvider from '@openworklabs/filecoin-wallet-provider'

import WalletView from '.'
import composeMockAppTree from '../../test-utils/composeMockAppTree'

jest.mock('@openworklabs/filecoin-wallet-provider')

describe('WalletView', () => {
  afterEach(cleanup)
  test.only('it renders correctly', () => {
    const { Tree } = composeMockAppTree('postOnboard')
    const { container } = render(
      <Tree>
        <WalletView />
      </Tree>
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
