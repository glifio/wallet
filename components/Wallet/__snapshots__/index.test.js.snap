// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletView it renders correctly 1`] = `
<div
  class="SidebarLayout__Wrapper-gwfrwp-0 fSFMLo"
>
  <div
    class="SidebarLayout__Sidebar-gwfrwp-1 VvNzW"
    height="100vh"
  >
    <div
      class="Box-tuskkk-0 gQWkUU"
      display="flex"
      width="100%"
    >
      <button
        class="NetworkSwitcherGlyph__NetworkSwitcherButton-svtr5s-0 ecYeVS"
        display="flex"
        font-family="RT-Alias-Grotesk"
        font-size="2"
        font-weight="1"
      >
        Testnet
      </button>
      <button
        class="NetworkSwitcherGlyph__NetworkSwitcherButton-svtr5s-0 gWYxtz"
        display="flex"
        font-family="RT-Alias-Grotesk"
        font-size="2"
        font-weight="1"
      >
        Mainnet
      </button>
    </div>
    <div
      class="Box-tuskkk-0 gcUhlS"
      color="purple"
      display="flex"
      height="300px"
      width="300px"
    >
      <div
        class="Box-tuskkk-0 hyonjk"
        color="card.account.color"
        display="flex"
      >
        <div
          class="Box-tuskkk-0 jFCwkq"
          color="card.account.color"
          display="flex"
          size="6"
        >
          <h3
            class="Glyph__GlyphText-sc-1xcgbq7-0 fuRHwt"
          >
            Ac
          </h3>
        </div>
        <p
          class="Typography__TextBase-sc-1uzug1e-2 LZLLr"
          font-family="RT-Alias-Grotesk"
          font-size="3"
          font-weight="400"
        >
          Account
        </p>
      </div>
      <div
        class="Box-tuskkk-0 jGWrYN"
        color="card.account.color"
      >
        <h1
          class="Typography__H1Base-sc-1uzug1e-0 jFAtte"
          font-family="RT-Alias-Medium"
          font-size="6"
          font-weight="700"
        >
          Prime
        </h1>
        <div
          class="Box-tuskkk-0 cDDaVE"
          display="flex"
        >
          <h2
            class="Typography__H2Base-sc-1uzug1e-1 dXRWZg"
            font-family="RT-Alias-Grotesk"
            font-size="5"
            font-weight="1"
          >
            t1z22 ... 6wgi
          </h2>
          <div
            class="Box-tuskkk-0 AccountCard__CopyAddress-sc-56bj9g-1 iWpbSq"
            display="flex"
          >
            <button
              class="IconButtons__IconButtonBase-sc-1st7bxm-0 dibZhS"
              display="inline-block"
            >
              <svg
                class="Icons__IconBase-jyp17t-0 hfFiWB"
                fill="none"
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="#1A0066"
                  height="12"
                  width="12"
                  x="8"
                  y="4"
                />
                <rect
                  fill="#E0D7FE"
                  height="12"
                  width="12"
                  x="4"
                  y="8"
                />
              </svg>
            </button>
            <h4
              class="Typography__H4Base-sc-1uzug1e-3 gfSgmy AccountCard__LabelCopy-sc-56bj9g-0 jsbuxv"
              font-family="RT-Alias-Grotesk"
              font-size="2"
              font-weight="400"
            >
              Copy
            </h4>
          </div>
        </div>
      </div>
      <div
        class="Box-tuskkk-0 dsKMJn"
        display="flex"
      >
        <button
          class="BaseButton-sc-1iyr4rl-0 bjpmLG"
          font-size="3"
          height="max-content"
          type="button"
        >
          Switch
        </button>
      </div>
    </div>
    <div
      class="Box-tuskkk-0 kaJSFe"
      display="flex"
      width="300px"
    >
      <div
        class="Box-tuskkk-0 gUAjrr"
        display="flex"
      >
        <h4
          class="Typography__H4Base-sc-1uzug1e-3 dvsHEz"
          font-family="RT-Alias-Grotesk"
          font-size="2"
          font-weight="400"
        >
          Balance
        </h4>
        <div
          class="Box-tuskkk-0 gUAjrr"
          display="flex"
        >
          <h4
            class="Typography__H4Base-sc-1uzug1e-3 bUeWBY ApproximationToggleBtn-sc-1lufoap-0 enNeUo"
            color="core.black"
            font-family="RT-Alias-Grotesk"
            font-size="2"
            font-weight="400"
          >
            Approx.
          </h4>
          <div
            class="Box-tuskkk-0 jjDyck"
            width="2"
          />
          <h4
            class="Typography__H4Base-sc-1uzug1e-3 AYXIM ApproximationToggleBtn-sc-1lufoap-0 enNeUo"
            color="core.darkgray"
            font-family="RT-Alias-Grotesk"
            font-size="2"
            font-weight="400"
          >
            Exact
          </h4>
        </div>
      </div>
      <div
        class="Box-tuskkk-0 fZmrbs"
        overflow="hidden"
      >
        <h2
          class="Typography__H2Base-sc-1uzug1e-1 kMGPcr BalanceCard___StyledNum-nx0aiw-0 eMLohY"
          color="card.balance.color"
          font-family="RT-Alias-Medium"
          font-size="6"
          font-weight="700"
        >
          &gt; 999.9B
           FIL
        </h2>
        <h2
          class="Typography__H2Base-sc-1uzug1e-1 jNkuEy"
          color="core.darkgray"
          font-family="RT-Alias-Medium"
          font-size="5"
          font-weight="700"
        >
          5 USD
        </h2>
      </div>
      <div
        class="Box-tuskkk-0 MrVvn"
        display="flex"
      >
        <button
          class="BaseButton-sc-1iyr4rl-0 gIInmt"
          font-size="3"
          type="button"
          width="120"
        >
          Receive
        </button>
        <button
          class="BaseButton-sc-1iyr4rl-0 bpdjDz"
          font-size="3"
          type="button"
          width="120"
        >
          Send
        </button>
      </div>
    </div>
    <button
      class="BaseButton-sc-1iyr4rl-0 Wallet___StyledButtonLogout-enfiyv-0 dMwnLp"
      display="flex"
      font-size="3"
      width="100%"
    >
      Logout
    </button>
  </div>
  <div
    class="SidebarLayout__Content-gwfrwp-2 eAcbru"
  >
    <div
      class="Box-tuskkk-0 kgyUWF"
      width="100%"
    >
      <div
        class="Box-tuskkk-0 cgmsIS"
        display="flex"
      >
        <div
          class="Box-tuskkk-0 gKEPob"
          color="core.primary"
          display="flex"
          size="6"
        >
          <h3
            class="Glyph__GlyphText-sc-1xcgbq7-0 fuRHwt"
          >
            Tx
          </h3>
        </div>
        <p
          class="Typography__TextBase-sc-1uzug1e-2 kgUsKE"
          color="core.primary"
          font-family="RT-Alias-Grotesk"
          font-size="3"
          font-weight="400"
        >
          Transaction History
        </p>
      </div>
      <div
        class="Box-tuskkk-0 kCebST"
        display="flex"
      >
        <span
          class="LoaderGlyph__LoaderGlyphParent-fi66in-0 dqjuhj"
        >
          <span
            class="LoaderGlyph__LoaderGlyphChild-fi66in-1 guWxTG"
          />
        </span>
        <h4
          class="Typography__H4Base-sc-1uzug1e-3 kTkIxR"
          font-family="RT-Alias-Grotesk"
          font-size="2"
          font-weight="400"
        >
          Loading...
        </h4>
      </div>
    </div>
  </div>
</div>
`;
